rust/psibase_macros/psibase-macros-lib/src/service_macro/actions.rs

```diff
@@ -190,3 +190,121 @@ pub fn process_action_schema(
+                // println!("pre_action_attr {:?}", pre_action_attr);
+                pre_action_info.exists = true;
+                pre_action_info.fn_name = Some(f.sig.ident.clone());
+                //TODO: decode attrs and get `exclude` arg
+                // println!("1: processing pre_action_attr args");
+                let attr_args_ts = match pre_action_attr.meta.clone() {
+                    Meta::List(args) => {
+                        // println!("List: {:?}", args);
+                        args.tokens
+                    }
+                    Meta::Path(args) => abort!(
+                        args,
+                        "Invalid pre_action attributes: expected list; got path."
+                    ),
+                    Meta::NameValue(args) => {
+                        // println!("NameValue: {:?}", args);
+                        args.value.to_token_stream()
+                    }
+                };
+                let attr_args = match NestedMeta::parse_meta_list(attr_args_ts.clone()) {
+                    Ok(v) => {
+                        // println!("v in match: {:#?}", v);
+                        v
+                    }
+                    Err(e) => {
+                        abort!(
+                            attr_args_ts,
+                            format!("Invalid pre_action arguments: {}", e.to_string())
+                        );
+                    }
+                };
+                // println!("2: parsing attr args into options...");
+                // println!("attr_args: {:#?}", attr_args);
+
+                let options: PreActionOptions =
+                    match PreActionOptions::from_list(&attr_args.clone()) {
+                        Ok(val) => val,
+                        Err(err) => {
+                            abort!(
+                                attr_args_ts,
+                                format!("Invalid pre_action arguments: {}", err.to_string())
+                            );
+                        }
+                    };
+                // println!("pre-action options: {:?}", options);
+                pre_action_info.exclude = options.exclude;
+
+                if let Some(pre_action_pos) = f.attrs.iter().position(is_pre_action_attr) {
+                    f.attrs.remove(pre_action_pos);
+                }
+            }
+        }
+    }
+}
+
+pub fn add_pre_action_call(pre_action_info: &PreAction, f: &mut ItemFn) {
+    // println!("adding check_init to {}", f.sig.ident.to_string());
+    if pre_action_info
+        .exclude
+        .to_strings()
+        .contains(&f.sig.ident.to_string())
+    {
+        // println!(
+        //     "{} is in exclude list; skipping adding pre_action call",
+        //     f.sig.ident.to_string()
+        // );
+        return;
+    }
+    let fn_name = pre_action_info.fn_name.clone();
+    let new_line_ts = quote! {
+        #fn_name();
```

Comment:
Just double checking that this works even though `fn_name` is an `Option<Ident>`?

Reply:
It does, but I did not intend to be relying on that, so I've cleaned up the code.